src/ir/effects.h

```diff
@@ -422,14 +424,25 @@ class EffectAnalyzer {
         return;
       }
 
+      if (curr->isReturn) {
+        parent.branchesOut = true;
+      }
+
+      if (parent.funcEffectsMap) {
+        auto iter = parent.funcEffectsMap->find(curr->target);
+        if (iter != parent.funcEffectsMap->end()) {
+          // We have effect information for this call target, and can just use
+          // that.
+          parent.mergeIn(iter->second);
+          return;
+        }
+      }
```

Comment:
> But this may not be so when analyses run in parallel.

Yes, atm this runs in parallel, and each function doesn't look any deeper than itself.

We could do a transitive closure operation afterwards. But I'm not sure it's worth it. This simple version already should handle all the real-world cases I can think of (with enough cycles of the optimizer, since each time it makes progress - there are no "cycles" to worry about here, as a cycle would be a potential infinite recursion anyhow, that we can't remove).

Reply:
What if the target effect is in the process of being generated? Is it safe to merge it here while it hasn't been done?

Reply:
`funcEffectsMap` is empty during the parallel computation (we clear it before, and only write to it after the parallel part). So we'd never reach this code path during the parallel computation.